EIPS/eip-7612.md

```diff
@@ -0,0 +1,181 @@
+        cskey[31] = CODE_SIZE_LEAF_KEY
+        tree.set(cskey, len(account.code))
+
+# Reads an account from the verkle tree
+def verkle_get_account(tree: VerkleTree, key: Bytes32) -> Optional[Account]:
+    version_leaf = tree.get(key)
+    if version_leaf is not None:
+        balancekey = key
+        balancekey[31] = BALANCE_LEAF_KEY
+        balance = tree.get(balancekey, account.balance)
+        noncekey = key
+        noncekey[31] = NONCE_LEAF_KEY
+        nonce = tree.get(noncekey)
+        ckkey = key
+        ckkey[31] = CODE_KECCAK_LEAF_KEY
+        ck = tree.get(ckkey)
+        cskey = key
+        cskey[31] = CODE_SIZE_LEAF_KEY
+        cs = tree.set(cskey)
+        account = Account(0, balance, nonce, ck, cs)
+
+    return account
+```
+
+### Changes to the execution spec:
+
+In the execution spec, modify the `State` class as such:
+
+```python3
+@dataclass
+class State:
+    """
+    Contains all information that is preserved between transactions.
+    """
+
+    _main_trie: Trie[Address, Optional[Account]] = field(
+        default_factory=lambda: Trie(secured=True, default=None)
+    )
+    _storage_tries: Dict[Address, Trie[Bytes, U256]] = field(
+        default_factory=dict
+    )
+    _snapshots: List[
+        Tuple[
+            Trie[Address, Optional[Account]], Dict[Address, Trie[Bytes, U256]]
+        ]
+    ] = field(default_factory=list)
+    _created_accounts: Set[Address] = field(default_factory=set)
+
+    # Added in this EIP
+    _overlay_tree: VerkleTree[Address, Bytes32]
+```
+
+And the state access functions are modified as such:
+
+```python3
+def get_account_optional(state: State, address: Address) -> Optional[Account]:
+    account = verkle_get_account(state._overlay_tree, get_tree_key_for_version(addr))
+    if account is not None:
+        return account
+    
+    return trie_get(state._main_trie, address)
+
+def set_account(state: State, address: Address, account: Optional[Account]) -> None:
+    verkle_set_account(state._overlay_tree, get_tree_key_for_nonce(addr), account)
+
+def get_storage(state: State, address: Address, key: Bytes) -> U256:
+    value = state._overlay_tree.get(get_tree_key_for_storage_slot(addr, slot))
+    if value is not None:
+        return value
+        
+    trie = state._storage_tries.get(address)
+    if trie is None:
+        return U256(0)
+
+    value = trie_get(trie, key)
+
+    assert isinstance(value, U256)
+    return value
+
+def set_storage(
+    state: State, address: Address, key: Bytes, value: U256
+) -> None:
+    state._overlay_tree.set(get_tree_key_for_storage_slot(addr, slot), value)
+```
+
+## Rationale
+
+This approach doesn't convert the state, which is left to a subsequent EIP. This is meant as a stopgap in case we decide to push the conversion itself to a later time. It has the advantage of simplicity, which means that the verge fork could happen at the same time as other, simpler EIPs. It also requires no change at the consensus layer.
```

Comment:
```suggestion
This approach doesn't convert the state, which is left to a subsequent EIP. This is meant as a stopgap in case we decide to push the conversion itself to a later time. It has the advantage of simplicity, which means that the Verge fork could happen at the same time as other, simpler EIPs. It also requires no change at the consensus layer.
```